File: it-agents/skills/code-architect/references/code-examples/ruby/decorator.rb
# frozen_string_literal: true

# Decorator — attach responsibilities by wrapping objects at runtime.

# Core component.
class PlainMessage
  def initialize(text) = @text = text
  def render = @text
end

# Base decorator forwards to the wrapped component.
class MessageDecorator
  def initialize(component) = @component = component
  def render = @component.render
end

# Concrete decorators add behavior around the wrapped result.
class Uppercase < MessageDecorator
  def render = @component.render.upcase
end

class Exclaim < MessageDecorator
  def render = "#{@component.render}!"
end

class Bracketed < MessageDecorator
  def render = "[#{@component.render}]"
end

if __FILE__ == $PROGRAM_NAME
  base = PlainMessage.new("hello")
  puts base.render

  # Decorators stack in any order, each wrapping the previous.
  decorated = Bracketed.new(Exclaim.new(Uppercase.new(base)))
  puts decorated.render
end

File: it-agents/skills/code-architect/references/code-examples/typescript/Decorator.ts
// Decorator — attach responsibilities by wrapping objects at runtime.

interface TextSource {
  read(): string;
}

class PlainText implements TextSource {
  constructor(private readonly content: string) {}
  read(): string {
    return this.content;
  }
}

// Base decorator forwards to the wrapped component.
abstract class TextDecorator implements TextSource {
  constructor(protected readonly wrapped: TextSource) {}
  abstract read(): string;
}

class UpperCaseDecorator extends TextDecorator {
  read(): string {
    return this.wrapped.read().toUpperCase();
  }
}

class BracketDecorator extends TextDecorator {
  read(): string {
    return `[${this.wrapped.read()}]`;
  }
}

function demo(): void {
  let source: TextSource = new PlainText("hello");
  source = new UpperCaseDecorator(source);
  source = new BracketDecorator(source);
  console.log(source.read()); // [HELLO]
}

demo();
